### ISSUE TITLE
Patch with nan values

### ISSUE THREAD

**Issue by mrocklin** [bokeh/bokeh#7525]
Is it possible to patch values in a ColumnDataSource with nan's?

It is difficult for Bokeh to json-serialize NaN values.  The usual fix is to convert to numpy arrays.  However it's not clear how to send numpy arrays when using `ColumnDataSource.patch`.  

A naive approach to just call `np.asarray` on the list of patches fails because the dtype is float, and bokeh expects integers.

```python
>>> source.patch({'x': np.asarray(patches)})

  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/tornado/gen.py", line 1069, in run
    yielded = self.gen.send(value)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/server/session.py", line 51, in _needs_document_lock_wrapper
    result = yield yield_for_all_futures(func(self, *args, **kwargs))
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/server/session.py", line 154, in with_document_locked
    return func(*args, **kwargs)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 1021, in wrapper
    return doc._with_self_as_curdoc(invoke)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 1007, in _with_self_as_curdoc
    return f()
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 1020, in invoke
    return f(*args, **kwargs)
  File "/home/mrocklin/workspace/distributed/distributed/bokeh/scheduler.py", line 704, in update
    'y': np.asarray(L)})
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/models/sources.py", line 626, in patch
    raise ValueError("Invalid patch index: %s" % ind)
ValueError: Invalid patch index: 1683.0
```

**Comment by mrocklin** [comment]
I also tried using a struct dtype

                L = np.asarray(L, dtype=[('a', 'i8'), ('b', 'f8')])

This gets a bit further, expecially if I change a couple `isinstance(ind, int)` checks to `isinstance(ind, numbers.Integral)`, but in the end this fails on JSON serialization

<details>

```python
tornado.application - ERROR - Future <tornado.concurrent.Future object at 0x7f01952fa198> exception was never retrieved: Traceback (most recent call last):
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/tornado/gen.py", line 1069, in run
    yielded = self.gen.send(value)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/server/session.py", line 51, in _needs_document_lock_wrapper
    result = yield yield_for_all_futures(func(self, *args, **kwargs))
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/server/session.py", line 154, in with_document_locked
    return func(*args, **kwargs)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 1021, in wrapper
    return doc._with_self_as_curdoc(invoke)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 1007, in _with_self_as_curdoc
    return f()
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 1020, in invoke
    return f(*args, **kwargs)
  File "/home/mrocklin/workspace/distributed/distributed/bokeh/scheduler.py", line 704, in update
    'y': np.asarray(L)})
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/models/sources.py", line 629, in patch
    self.data._patch(self.document, self, patches, setter)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/core/property/containers.py", line 441, in _patch
    hint=ColumnsPatchedEvent(doc, source, patches, setter))
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/core/property/containers.py", line 111, in _notify_owners
    descriptor._notify_mutated(owner, old, hint=hint)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/core/property/descriptors.py", line 833, in _notify_mutated
    self._real_set(obj, old, value, hint=hint)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/core/property/descriptors.py", line 796, in _real_set
    self._trigger(obj, old, value, hint=hint, setter=setter)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/core/property/descriptors.py", line 873, in _trigger
    obj.trigger(self.name, old, value, hint, setter)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/model.py", line 543, in trigger
    super(Model, self).trigger(attr, old, new, hint=hint, setter=setter)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/util/callback_manager.py", line 118, in trigger
    self._document._notify_change(self, attr, old, new, hint, setter, invoke)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 909, in _notify_change
    self._trigger_on_change(event)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 994, in _trigger_on_change
    self._with_self_as_curdoc(invoke_callbacks)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 1007, in _with_self_as_curdoc
    return f()
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 993, in invoke_callbacks
    cb(event)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/document.py", line 587, in <lambda>
    self._callbacks[receiver] = lambda event: event.dispatch(receiver)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/events.py", line 195, in dispatch
    super(ModelChangedEvent, self).dispatch(receiver)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/document/events.py", line 77, in dispatch
    receiver._document_patched(self)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/server/session.py", line 186, in _document_patched
    self._pending_writes.append(connection.send_patch_document(event))
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/server/connection.py", line 42, in send_patch_document
    msg = self.protocol.create('PATCH-DOC', [event])
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/protocol/__init__.py", line 46, in create
    return self._messages[msgtype].create(*args, **kwargs)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/protocol/messages/patch_doc.py", line 55, in create
    patch_json, buffers = process_document_events(events, use_buffers)
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/protocol/messages/patch_doc.py", line 98, in process_document_events
    return serialize_json(json), buffers if use_buffers else []
  File "/home/mrocklin/anaconda/lib/python3.6/site-packages/bokeh/core/json_encoder.py", line 213, in serialize_json
    return json.dumps(obj, cls=BokehJSONEncoder, allow_nan=False, indent=indent, separators=separators, sort_keys=True, **kwargs)
  File "/home/mrocklin/anaconda/lib/python3.6/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/home/mrocklin/anaconda/lib/python3.6/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/mrocklin/anaconda/lib/python3.6/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
ValueError: Out of range float values are not JSON compliant
```

</details>

**Comment by bryevdv** [comment]
> It is difficult for Bokeh to json-serialize NaN values. 

That's because JSON itself does not serialize `NaN` values (cc @BrendanEich)

There's not going to be a great solution to this, perhaps we can detect real Python `NaN` values here:

https://github.com/bokeh/bokeh/blob/master/bokeh/core/property/containers.py#L399

 and convert them to e.g. `"NaN"`, which I believe should by coincidence result in a real JS `NaN` in the necessary places. Under the assumption that patches are typically very small (that is the intended use case) then the added overhead may not be prohibitive.

**Comment by bryevdv** [comment]
As an update, patching the string `"nan"` does work. This is a happy (?) accident of JavaScript's nightmare implicit conversions. Presumably any junk string would also work. 

I am not especially happy about this, but it seems to be a somewhat obscure corner case with a documented workaround here now, so I am going to close this. If there is renewed or increased interest later, perhaps we can revisit (I don't think there is a trivial fix)

**Comment by chesucr** [comment]
In relation to this I have asked for an algorythm to group all repeated values [**here**](https://stackoverflow.com/a/56274369/4891717) to check if it would improve the performance. The vectorized version that Quang Hoang is suggesting takes longer than the other one.

I had to make a small correction in the algorythm to make it work with integer indices, instead of `ser.index` I needed to use the integer values:

    indices = ser.index.tolist()       # Int64 indices are casted to int

The conslusion is that I do not really know if it improves something because I do not know how to measure it. So if anybody is interested in doing it let me know. My eye cannot see any difference
